File: src/code-samples/langchain/agents-harness.ts
// Agents docs: harness examples (execution environment, context management, middleware).

// :remove-start:
import { tool } from "langchain";
import * as z from "zod";

var search = tool(({ query }) => `Results for: ${query}`, {
  name: "search",
  description: "Search for information",
  schema: z.object({ query: z.string() }),
});
// :remove-end:

// :snippet-start: agents-execution-environment-js
import { createAgent } from "langchain";
import { createFilesystemMiddleware, StateBackend } from "deepagents";

var agent = createAgent({
  model: "anthropic:claude-sonnet-4-6",
  tools: [search],
  middleware: [createFilesystemMiddleware({ backend: new StateBackend() })],
});
// :snippet-end:

// :remove-start:
const executionEnvironmentAgent = agent;
// :remove-end:

// :snippet-start: agents-context-management-js
import { createAgent } from "langchain";
import {
  StateBackend,
  createFilesystemMiddleware,
  createSkillsMiddleware,
  createSummarizationMiddleware,
} from "deepagents";

var backend = new StateBackend();
const model = "anthropic:claude-sonnet-4-6";

var agent = createAgent({
  model,
  tools: [search],
  middleware: [
    createFilesystemMiddleware({ backend }),
    createSummarizationMiddleware({ model, backend }),
    createSkillsMiddleware({ backend, sources: ["./skills/"] }),
  ],
});
// :snippet-end:

// :remove-start:
const contextManagementAgent = agent;
// :remove-end:

// :snippet-start: agents-fault-tolerance-js
import {
  createAgent,
  modelRetryMiddleware,
  tool,
  toolRetryMiddleware,
} from "langchain";
import * as z from "zod";

var search = tool(({ query }) => `Search results for: ${query}`, {
  name: "search",
  description: "Search for a query and return a short summary.",
  schema: z.object({ query: z.string() }),
});

var agent = createAgent({
  model: "anthropic:claude-sonnet-4-6",
  tools: [search],
  middleware: [
    modelRetryMiddleware({ maxRetries: 3 }),
    toolRetryMiddleware({ maxRetries: 2 }),
  ],
});
// :snippet-end:

// :remove-start:
const faultToleranceAgent = agent;
// :remove-end:

// :snippet-start: agents-guardrails-js
import { createAgent, piiMiddleware, tool } from "langchain";
import * as z from "zod";

var search = tool(({ query }) => `Search results for: ${query}`, {
  name: "search",
  description: "Search for a query and return a short summary.",
  schema: z.object({ query: z.string() }),
});

var agent = createAgent({
  model: "anthropic:claude-sonnet-4-6",
  tools: [search],
  middleware: [piiMiddleware("email")],
});
// :snippet-end:

// :remove-start:
const guardrailsAgent = agent;
// :remove-end:

// :snippet-start: agents-planning-delegation-js
import { createAgent, todoListMiddleware, tool } from "langchain";
import {
  createFilesystemMiddleware,
  createSubAgentMiddleware,
  StateBackend,
} from "deepagents";
import * as z from "zod";

var search = tool(({ query }) => `Search results for: ${query}`, {
  name: "search",
  description: "Search for a query and return a short summary.",
  schema: z.object({ query: z.string() }),
});

var backend = new StateBackend();

var agent = createAgent({
  model: "anthropic:claude-sonnet-4-6",
  tools: [search],
  middleware: [
    createFilesystemMiddleware({ backend }),
    todoListMiddleware(),
    createSubAgentMiddleware({
      defaultModel: "anthropic:claude-sonnet-4-6",
      defaultTools: [],
      subagents: [
        {
          name: "researcher",
          description: "Searches and returns a structured summary.",
          systemPrompt:
            "Use the search tool to research the question and summarize key points.",
          tools: [search],
          model: "anthropic:claude-sonnet-4-6",
          middleware: [],
        },
      ],
    }),
  ],
});
// :snippet-end:

// :remove-start:
const planningDelegationAgent = agent;
// :remove-end:

// :snippet-start: agents-steering-js
import { createAgent, humanInTheLoopMiddleware, tool } from "langchain";
import * as z from "zod";

var search = tool(({ query }) => `Search results for: ${query}`, {
  name: "search",
  description: "Search for a query and return a short summary.",
  schema: z.object({ query: z.string() }),
});

var agent = createAgent({
  model: "anthropic:claude-sonnet-4-6",
  tools: [search],
  middleware: [humanInTheLoopMiddleware({ interruptOn: { writeFile: true } })],
});
// :snippet-end:

// :remove-start:
import { AIMessage } from "@langchain/core/messages";

const steeringAgent = agent;

function assistantText(result: { messages: unknown[] }): string {
  const last = result.messages[result.messages.length - 1];
  if (!(last instanceof AIMessage)) {
    throw new Error("expected final assistant message");
  }
  const blocks = last.contentBlocks ?? [];
  const textBlock = blocks.find((b) => b.type === "text");
  if (textBlock && textBlock.type === "text") {
    return textBlock.text;
  }
  return typeof last.content === "string" ? last.content : "";
}

async function assertAgentInvokes(
  testAgent: typeof agent,
  label: string,
): Promise<void> {
  const result = await testAgent.invoke({
    messages: [
      {
        role: "user",
        content: "Use the search tool to look up LangChain agents.",
      },
    ],
  });
  if (!assistantText(result).trim()) {
    throw new Error(`${label}: expected non-empty assistant reply`);
  }
  console.log(`✓ ${label} invokes and returns a response`);
}

async function main() {
  await assertAgentInvokes(
    executionEnvironmentAgent,
    "agents execution environment",
  );
  await assertAgentInvokes(contextManagementAgent, "agents context management");
  await assertAgentInvokes(faultToleranceAgent, "agents fault tolerance");
  await assertAgentInvokes(guardrailsAgent, "agents guardrails");
  await assertAgentInvokes(
    planningDelegationAgent,
    "agents planning/delegation",
  );
  await assertAgentInvokes(steeringAgent, "agents steering");
}

main();
// :remove-end:

File: src/code-samples/langchain/agents-harness.py
"""Agents docs: harness examples (execution environment, context management, middleware)."""

# :remove-start:
from langchain.tools import tool


@tool
def search(query: str) -> str:
    """Search for information."""
    return f"Results for: {query}"
# :remove-end:

# :snippet-start: agents-execution-environment-py
from langchain.agents import create_agent
from deepagents.backends import StateBackend
from deepagents.middleware import FilesystemMiddleware

agent = create_agent(
    model="anthropic:claude-sonnet-4-6",
    tools=[search],
    middleware=[FilesystemMiddleware(backend=StateBackend())],
)
# :snippet-end:

# :remove-start:
_execution_environment_agent = agent
# :remove-end:

# :snippet-start: agents-context-management-py
from deepagents.backends import StateBackend
from deepagents.middleware import FilesystemMiddleware, MemoryMiddleware, SkillsMiddleware, SummarizationMiddleware

backend = StateBackend()
model = "anthropic:claude-sonnet-4-6"

agent = create_agent(
    model=model,
    tools=[search],
    middleware=[
        FilesystemMiddleware(backend=backend),
        SummarizationMiddleware(model=model, backend=backend),
        MemoryMiddleware(backend=backend, sources=["./AGENTS.md"]),
        SkillsMiddleware(backend=backend, sources=["./skills/"]),
    ],
)
# :snippet-end:

# :remove-start:
_context_management_agent = agent
# :remove-end:

# :snippet-start: agents-fault-tolerance-py
from langchain.agents import create_agent
from langchain.agents.middleware import ModelRetryMiddleware, ToolRetryMiddleware
from langchain.tools import tool


@tool
def search(query: str) -> str:
    """Search for a query and return a short summary."""
    return f"Search results for: {query}"


agent = create_agent(
    model="anthropic:claude-sonnet-4-6",
    tools=[search],
    middleware=[
        ModelRetryMiddleware(max_retries=3),
        ToolRetryMiddleware(max_retries=2),
    ],
)
# :snippet-end:

# :remove-start:
_fault_tolerance_agent = agent
# :remove-end:

# :snippet-start: agents-guardrails-py
from langchain.agents import create_agent
from langchain.agents.middleware import PIIMiddleware
from langchain.tools import tool


@tool
def search(query: str) -> str:
    """Search for a query and return a short summary."""
    return f"Search results for: {query}"


agent = create_agent(
    model="anthropic:claude-sonnet-4-6",
    tools=[search],
    middleware=[PIIMiddleware("email")],
)
# :snippet-end:

# :remove-start:
_guardrails_agent = agent
# :remove-end:

# :snippet-start: agents-planning-delegation-py
from deepagents.backends import StateBackend
from deepagents.middleware import FilesystemMiddleware
from deepagents.middleware.subagents import SubAgentMiddleware
from langchain.agents import create_agent
from langchain.agents.middleware import TodoListMiddleware
from langchain.tools import tool


@tool
def search(query: str) -> str:
    """Search for a query and return a short summary."""
    return f"Search results for: {query}"


backend = StateBackend()

agent = create_agent(
    model="anthropic:claude-sonnet-4-6",
    tools=[search],
    middleware=[
        FilesystemMiddleware(backend=backend),
        TodoListMiddleware(),
        SubAgentMiddleware(
            backend=backend,
            subagents=[
                {
                    "name": "researcher",
                    "description": "Searches and returns a structured summary.",
                    "system_prompt": "Use the search tool to research the question and summarize key points.",
                    "tools": [search],
                    "model": "anthropic:claude-sonnet-4-6",
                    "middleware": [],
                }
            ],
        ),
    ],
)
# :snippet-end:

# :remove-start:
_planning_delegation_agent = agent
# :remove-end:

# :snippet-start: agents-steering-py
from langchain.agents import create_agent
from langchain.agents.middleware import HumanInTheLoopMiddleware
from langchain.tools import tool


@tool
def search(query: str) -> str:
    """Search for a query and return a short summary."""
    return f"Search results for: {query}"


agent = create_agent(
    model="anthropic:claude-sonnet-4-6",
    tools=[search],
    middleware=[HumanInTheLoopMiddleware(interrupt_on={"write_file": True})],
)
# :snippet-end:

# :remove-start:
_steering_agent = agent


def _assistant_text(result: dict) -> str:
    last = result["messages"][-1]
    if last.content_blocks:
        return last.content_blocks[0]["text"]
    return str(last.content)


def _assert_agent_invokes(agent, label: str) -> None:
    result = agent.invoke(
        {
            "messages": [
                {
                    "role": "user",
                    "content": "Use the search tool to look up LangChain agents.",
                }
            ]
        }
    )
    text = _assistant_text(result)
    assert text.strip(), f"{label}: expected non-empty assistant reply"
    print(f"✓ {label} invokes and returns a response")


if __name__ == "__main__":
    _assert_agent_invokes(_execution_environment_agent, "agents execution environment")
    _assert_agent_invokes(_context_management_agent, "agents context management")
    _assert_agent_invokes(_fault_tolerance_agent, "agents fault tolerance")
    _assert_agent_invokes(_guardrails_agent, "agents guardrails")
    _assert_agent_invokes(_planning_delegation_agent, "agents planning/delegation")
    _assert_agent_invokes(_steering_agent, "agents steering")
# :remove-end:
